jina/serve/instrumentation/__init__.py

```diff
@@ -0,0 +1,123 @@
+from typing import Any, Optional
+
+
+class InstrumentationMixin:
+    '''Instrumentation mixin for OpenTelemetery Tracing and Metrics handling'''
+
+    def _setup_instrumentation(
+        self,
+        name: str,
+        tracing: Optional[bool] = False,
+        span_exporter_host: Optional[str] = '0.0.0.0',
+        span_exporter_port: Optional[int] = 6831,
+        metrics: Optional[bool] = False,
+        metrics_exporter_host: Optional[str] = '0.0.0.0',
+        metrics_exporter_port: Optional[int] = 6831,
+    ) -> None:
+
+        self.tracing = tracing
+        self.metrics = metrics
+
+        if tracing:
+            from opentelemetry import trace
+            from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import (
+                OTLPSpanExporter,
+            )
+            from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+            from opentelemetry.sdk.trace import TracerProvider
+            from opentelemetry.sdk.trace.export import BatchSpanProcessor
+
+            resource = Resource(attributes={SERVICE_NAME: name})
+            provider = TracerProvider(resource=resource)
+            processor = BatchSpanProcessor(
+                OTLPSpanExporter(
+                    endpoint=f'{span_exporter_host}:{span_exporter_port}', insecure=True
+                )
+            )
+            provider.add_span_processor(processor)
+            self.tracer_provider = provider
+            self.tracer = provider.get_tracer(name)
+        else:
+            from opentelemetry import trace
+
+            self.tracer_provider = trace.NoOpTracerProvider()
+            self.tracer = trace.NoOpTracer()
+
+        if metrics:
+            from opentelemetry import metrics as opentelmetry_metrics
+            from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import (
+                OTLPMetricExporter,
+            )
+            from opentelemetry.sdk.metrics import MeterProvider
+            from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
+            from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+
+            resource = Resource(attributes={SERVICE_NAME: name})
+
+            metric_reader = PeriodicExportingMetricReader(
+                OTLPMetricExporter(
+                    endpoint=f'{metrics_exporter_host}:{metrics_exporter_port}',
+                    insecure=True,
```

Comment:
can't we just have args: metric-exporter-kwargs ? (like in docker-kwargs parameter)
there are quite some interesting parameters left here

Reply:
We can but this will have to be provided in the next iteration because there are still quite a few details that we need to decide upon. This is currently not production ready.